infrastructure/smart-contracts/contracts/upgrades/pool-factory/PoolFactory3.sol

```diff
@@ -0,0 +1,875 @@
+            removeUserFromPool(msg.sender, pool);
+        }
+
+        emit UnstakeKeys(
+            msg.sender,
+            pool,
+            keyIds.length,
+            stakingPool.getStakedKeysCountForUser(msg.sender),
+            stakingPool.getStakedKeysCount()
+        );
+    }
+
+    /**
+     * @notice Stakes esXai in a pool.
+     * @param pool The address of the pool.
+     * @param amount The amount of esXai to stake.
+     */
+    function stakeEsXai(address pool, uint256 amount) external {
+        require(poolsCreatedViaFactory[pool], "27"); // Pool must be created via factory
+        require(failedKyc[msg.sender] == false, "38"); // Owner must not have failed kyc
+
+        Referee9(refereeAddress).stakeEsXai(pool, amount);
+        esXai(esXaiAddress).transferFrom(msg.sender, address(this), amount);
+        StakingPool stakingPool = StakingPool(pool);
+        stakingPool.stakeEsXai(msg.sender, amount);
+
+        associateUserWithPool(msg.sender, pool);
+
+        // Update total stake
+        if(!totalStakeCalculated[msg.sender]) {
+            _totalStakeByUser[msg.sender] = getTotalesXaiStakedByUser(msg.sender);
+            totalStakeCalculated[msg.sender] = true;
+        } else {
+            _totalStakeByUser[msg.sender] += amount;
+        }
+
+        emit StakeEsXai(
+            msg.sender,
+            pool,
+            amount,
+            stakingPool.getStakedAmounts(msg.sender),
+            Referee9(refereeAddress).stakedAmounts(pool)
+        );
+    }
+
+    /**
+     * @notice Unstakes esXai from a pool.
+     * @param pool The address of the pool.
+     * @param unstakeRequestIndex The index of the unstake request.
+     * @param amount The amount of esXai to unstake.
+     */
+    function unstakeEsXai(
+        address pool,
+        uint256 unstakeRequestIndex,
+        uint256 amount
+    ) external {
+        require(poolsCreatedViaFactory[pool], "28"); // Pool must be created via factory
+
+        esXai(esXaiAddress).transfer(msg.sender, amount);
+        Referee9(refereeAddress).unstakeEsXai(pool, amount);
+        StakingPool stakingPool = StakingPool(pool);
+        stakingPool.unstakeEsXai(msg.sender, unstakeRequestIndex, amount);
+
+        if (!stakingPool.isUserEngagedWithPool(msg.sender)) {
+            removeUserFromPool(msg.sender, pool);
+        }
+
+        // Update total stake
+        if(!totalStakeCalculated[msg.sender]) {
+            uint256 totalPools = interactedPoolsOfUser[msg.sender].length;
+            // We check the total pools interacted with by the user to avoid running out of gas
+            // This ensures we do not run this one time calculation until we are sure that it will not run out of gas
+            // This ensures a user can always unstake their esXai
+            if(totalPools < 150) {
+                _totalStakeByUser[msg.sender] = getTotalesXaiStakedByUser(msg.sender);
+                totalStakeCalculated[msg.sender] = true;
+            }
```

Comment:
We might want to add a public write function that can sync that value. Anybody should be able to sync any user wallet, could be us that sync all the suers before we actually have to worry about a user having staked in too many pools. Let me know what you think.

Reply:
I agree, done!